api/v1alpha1/reclaimspacecronjob_types.go

```diff
@@ -0,0 +1,135 @@
+)
+
+// ReclaimSpaceJobTemplateSpec describes the data a Job should have when created from a template
+type ReclaimSpaceJobTemplateSpec struct {
+	// Standard object's metadata of the jobs created from this template.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of the job.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +kubebuilder:validation:Required
+	Spec ReclaimSpaceJobSpec `json:"spec,omitempty"`
+}
+
+// ReclaimSpaceCronJobSpec defines the desired state of ReclaimSpaceJob
+type ReclaimSpaceCronJobSpec struct {
+	// The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern:=.+
+	Schedule string `json:"schedule"`
+
+	// Specifies how to treat concurrent executions of a Job.
+	// Valid values are:
+	// - "Forbid" (default): forbids concurrent runs, skipping next run if
+	//   previous run hasn't finished yet;
+	// - "Replace": cancels currently running job and replaces it
+	//   with a new one
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=Forbid;Replace
+	// +kubebuilder:default:=Forbid
+	ConcurrencyPolicy ConcurrencyPolicy `json:"concurrencyPolicy,omitempty"`
+
+	// This flag tells the controller to suspend subsequent executions, it does
+	// not apply to already started executions.  Defaults to false.
+	// +kubebuilder:validation:Optional
+	Suspend *bool `json:"suspend,omitempty"`
+
+	// Specifies the job that will be created when executing a CronJob.
+	// +kubebuilder:validation:Required
+	JobSpec ReclaimSpaceJobTemplateSpec `json:"jobTemplate"`
+
+	// The number of successful finished jobs to retain. Value must be non-negative integer.
+	// Defaults to 3.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Maximum=60
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:default:=3
+	SuccessfulJobsHistoryLimit *int32 `json:"successfulJobsHistoryLimit,omitempty"`
+
+	// The number of failed finished jobs to retain. Value must be non-negative integer.
+	// Defaults to 1.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Maximum=60
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:default:=1
+	FailedJobsHistoryLimit *int32 `json:"failedJobsHistoryLimit,omitempty"`
+}
+
+// ConcurrencyPolicy describes how the job will be handled.
+// Only one of the following concurrent policies may be specified.
+// If none of the following policies is specified, the default one
+// is ReplaceConcurrent.
+type ConcurrencyPolicy string
+
+const (
+	// ForbidConcurrent forbids concurrent runs, skipping next run if previous
+	// hasn't finished yet.
+	ForbidConcurrent ConcurrencyPolicy = "Forbid"
+
+	// ReplaceConcurrent cancels currently running job and replaces it with a new one.
```

Comment:
how we can cancel the currently running job and replace it with new one?

Reply:
> how we can cancel the currently running job and replace it with new one?

https://github.com/kubernetes-sigs/kubebuilder/blob/acdede50a58a1b9ac7fa66fc335bc810fe9ef6f6/docs/book/src/cronjob-tutorial/testdata/project/controllers/cronjob_controller.go#L441 
The entire controller will closely be based on kubebuilder tutorial for cronjobs

Reply:
yes, correct but how are you planning to cancel the currently running job?

Reply:
> yes, correct but how are you planning to cancel the currently running job?

delete reclaimspacejob CR, just like how k8s cronjob does.

Reply:
but you don't stop the RPC request already sent to the CSI driver to reclaim the space. if only deleting the CR's and creating a new one am fine with that.

Reply:
> but you don't stop the RPC request already sent to the CSI driver to reclaim the space. if only deleting the CR's and creating a new one am fine with that.

Yes, that's the case. :+1:

Reply:
Indeed the `ReclaimSpaceJob` that was started will get deleted (cancelled). Because there is no way to interrupt the running reclaim space operation, there is not much the controller can do about that.